### PR TITLE
Client close handling

### LINKED ISSUE #7: Expand Rabbit config options & channel recovery.
1. Rework the configuration and handling of RabbitMQ connections to allow the library to expose options to consumers.
2. Unexpected channel closure recovery.  If/when a channel is closed when a close wasn't explicitly requested, it should be recreated behind the scenes so consumers aren't halted unexpectedly.

### PR DESCRIPTION
**Pull request by Foxcapades**
### Changes

* Rework config class
* Rename QueueDispatcher to JobQueueDispatcher
* Rename QueueWorker to JobQueueExecutor
* Connections are no-longer re-created.  The initial implementation of this was problematic and making it work how it was originally intended will be a significant undertaking.

#### 1. Job Timeouts

The worker client now enforces a configurable job timeout.  By default the timeout is 15 minutes.

When setting up the connection to the rabbitmq instance the job timeout will be used to set the  consumer_timeout value for acknowledging messages on the work queue.

If a job exceeds this time it will be interrupted and killed.

#### 2. Failure Policies

To allow the overall process to be killed when/if the worker pool reaches a 'failed' state, the executor pool may now be provided with a failure policy which defines a condition or conditions in which the worker pool automatic recovery will halt and the pool will be shut down.

By default the worker pool will shut down if `2 * workerCount` unexpected rabbitmq client shutdowns happen within 2 seconds.

### Log Output Example

2 message acknowledgment timeout errors with a Failure Policy allowing for 1 total failure

```
[E] com.rabbitmq.client.AlreadyClosedException: channel is already closed due to channel error; protocol method: #method<channel.close>(reply-code=406, reply-text=PRECONDITION_FAILED - delivery acknowledgement on channel 1 timed out. Timeout value used: 5000 ms. This timeout value can be configured, see consumers doc guide to learn more, class-id=0, method-id=0)
[W] JobQueueExecutorPool - caught unexpected shutdown on consumer jobs-consumer-pool-1
[E] com.rabbitmq.client.AlreadyClosedException: channel is already closed due to channel error; protocol method: #method<channel.close>(reply-code=406, reply-text=PRECONDITION_FAILED - delivery acknowledgement on channel 1 timed out. Timeout value used: 5000 ms. This timeout value can be configured, see consumers doc guide to learn more, class-id=0, method-id=0)
[W] JobQueueExecutorPool - caught unexpected shutdown on consumer jobs-consumer-pool-2
[E] JobQueueExecutorPool - shutting down job queue executor: process exceeded 1 total failures
[I] JobQueueExecutor - closing connection to RabbitMQ
client-1 exited with code 0
```

### CODE REVIEW

File: src/main/kotlin/org/veupathdb/lib/rabbit/jobs/QueueWrapper.kt [hunk 1]
```diff
@@ -107,50 +90,46 @@ sealed class QueueWrapper {
    * Initializes the job dispatch queue if it is not already initialized.
    */
   protected open fun Channel.initDispatchQueue() {
-    queueDeclare(dispatchQueueName, true, false, false, emptyMap())
+    queueDeclare(
+      /* queue = */      dispatchQueueName,
+      /* durable = */    true,
+      /* exclusive = */  false,
+      /* autoDelete = */ false,
+      /* arguments = */  mapOf(
```
Comment (dmgaldi) [inline src/main/kotlin/org/veupathdb/lib/rabbit/jobs/QueueWrapper.kt:98]: Curious why the arg names are commented out here

File: src/main/kotlin/org/veupathdb/lib/rabbit/jobs/pools/exec/JobQueueExecutorPool.kt [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+      return if (threads.isTerminated) StopCode.Graceful else StopCode.Unknown
+
+    if (threads.awaitTermination(killAfter.inWholeMilliseconds, TimeUnit.MILLISECONDS))
+      return StopCode.Graceful
+
+    threads.shutdownNow()
+    silently(config.shutdownCB)
+    if (blockAfterKill)
+      threads.awaitTermination(243256, TimeUnit.DAYS)
+
```
Comment (dmgaldi) [inline src/main/kotlin/org/veupathdb/lib/rabbit/jobs/pools/exec/JobQueueExecutorPool.kt:108]: Is there any significance to this wait time, or is just effectively "wait indefinitely" and there's no other way to do that?

**Review by dmgaldi — APPROVED** [review]

File: src/main/kotlin/org/veupathdb/lib/rabbit/jobs/QueueWrapper.kt (hunk 1 above)
Comment (Foxcapades) [reply src/main/kotlin/org/veupathdb/lib/rabbit/jobs/QueueWrapper.kt:98]: The API is defined in Java not Kotlin, so arg name usage doesn't work the same

File: src/main/kotlin/org/veupathdb/lib/rabbit/jobs/pools/exec/JobQueueExecutorPool.kt (hunk 2 above)
Comment (Foxcapades) [reply src/main/kotlin/org/veupathdb/lib/rabbit/jobs/pools/exec/JobQueueExecutorPool.kt:108]: It is just wait indefinitely.  I forget what that specific number 'means', but that meaning was a joke anyway.  I will add a comment explaining it.

**Pull request merged**